pkg/gatewayserver/redis/registry.go

```diff
@@ -32,14 +36,66 @@ func (r *GatewayConnectionStatsRegistry) key(uid string) string {
 }
 
 // Set sets or clears the connection stats for a gateway.
-func (r *GatewayConnectionStatsRegistry) Set(ctx context.Context, ids ttnpb.GatewayIdentifiers, stats *ttnpb.GatewayConnectionStats) error {
+func (r *GatewayConnectionStatsRegistry) Set(ctx context.Context, ids ttnpb.GatewayIdentifiers, stats *ttnpb.GatewayConnectionStats, newTraffic io.Traffic) error {
 	uid := unique.ID(ctx, ids)
-	if stats == nil {
-		return r.Redis.Del(r.key(uid)).Err()
-	}
+	key := r.key(uid)
+
+	defer trace.StartRegion(ctx, "set gateway connection stats").End()
+
+	err := r.Redis.Watch(func(tx *redis.Tx) error {
+		stored := &ttnpb.GatewayConnectionStats{}
+		var pipelined func(redis.Pipeliner) error
+
+		// Delete if nil
+		if stats == nil {
+			pipelined = func(p redis.Pipeliner) error {
+				return p.Del(key).Err()
+			}
+		} else {
+			// Get current stats
+			err := ttnredis.GetProto(tx, key).ScanProto(stored)
+			if err != nil {
+				// Redis error, cannot proceed
+				if !errors.IsNotFound(err) {
+					return err
+				}
+
+				// No stats in store yet. Simply set.
+				pipelined = func(p redis.Pipeliner) error {
+					_, err := ttnredis.SetProto(p, key, stats, 0)
+					return err
+				}
+			} else {
+				// Existing stats are in stored. Update based on new traffic
+				if newTraffic.Up {
+					stored.LastUplinkReceivedAt = stats.LastUplinkReceivedAt
+					stored.UplinkCount = stats.UplinkCount
+				}
+				if newTraffic.Down {
+					stored.LastDownlinkReceivedAt = stats.LastDownlinkReceivedAt
+					stored.DownlinkCount = stats.DownlinkCount
+					stored.RoundTripTimes = stats.RoundTripTimes
+				}
+				if newTraffic.Status {
+					stored.LastStatus = stats.LastStatus
+					stored.LastStatusReceivedAt = stats.LastStatusReceivedAt
+				}
```

Comment:
Is there a reason _not_ to keep these in separate keys, while using the same proto?

Reply:
If we store in separate keys (and the stats are updated based on timestamp), then we have two options:

- For every stats update, first read all three keys from Redis (for up, down and status respectively), check if you have a newer timestamp and then update which are needed. (3 reads, <= 3 writes)
- Keep the timestamp of the last update as state, and blind update whichever stats have a more recent timestamp. This is not as safe as the first option (a blind write may be incorrect in this case), requires extra state for each `io.Connection`, but only does <= 3 writes.

Currently, storing on a single key needs 1 read, 1 write, but has more collisions with concurrent updates (e.g. from multiple server instances updating different stats).

Reply:
The thing is that uplink, downlink and status streams are statefully routed to instances, but separately, so that uplink, downlink and status may end up on different instances. You don’t have to account for multiple instances handling uplink, only for multiple instances handling uplink vs downlink.

You can therefore safely keep timestamps in memory for uplink, downlink and status separately.